src/python/pants/backend/experimental/visibility/register.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from dataclasses import dataclass
+from fnmatch import fnmatch
+from typing import Iterable, Mapping, Tuple, cast
+
+from pants.engine.internals.dep_rules import (
+    BuildFileDependencyRules,
+    BuildFileDependencyRulesImplementation,
+    BuildFileDependencyRulesImplementationRequest,
+    DependencyRule,
+    DependencyRuleAction,
+    DependencyRules,
+)
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    DependenciesRuleAction,
+    DependenciesRuleActionRequest,
+    FieldSet,
+    ValidatedDependencies,
+    ValidateDependenciesRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+
+
+class BuildFileVisibilityImplementationRequest(BuildFileDependencyRulesImplementationRequest):
+    pass
+
+
+class VisibilityValidateFieldSet(FieldSet):
+    required_fields = (Dependencies,)
+
+
+class VisibilityValidateDependenciesRequest(ValidateDependenciesRequest):
+    field_set_type = VisibilityValidateFieldSet
+
+
+@dataclass(frozen=True)
+class VisibilityRule(DependencyRule):
+    action: DependencyRuleAction
+    pattern: str
+
+    @classmethod
+    def parse(cls, rule: str) -> VisibilityRule:
+        if rule.startswith("!"):
+            action = DependencyRuleAction.DENY
+            pattern = rule[1:]
+        elif rule.startswith("?"):
+            action = DependencyRuleAction.WARN
+            pattern = rule[1:]
+        else:
+            action = DependencyRuleAction.ALLOW
+            pattern = rule
+        return cls(action, pattern)
+
+    def match(self, path: str, relpath: str) -> bool:
+        pattern = relpath if self.pattern == "." else self.pattern
+        if pattern.startswith("./"):
+            pattern = relpath + pattern[1:]
+        return fnmatch(path, pattern)
+
+
+VisibilityRules = Tuple[VisibilityRule, ...]
+
+
+@dataclass(frozen=True)
+class BuildFileVisibilityRules(BuildFileDependencyRules):
+    default: DependencyRuleAction
+    all: DependencyRules
+    targets: Mapping[str, DependencyRules]
+
+    @classmethod
+    def create(
+        cls,
+        default: DependencyRuleAction = DependencyRuleAction.ALLOW,
+        all: Iterable[str | DependencyRule] = (),
+        targets: Mapping[str, Iterable[str | DependencyRule]] = {},
+    ) -> BuildFileVisibilityRules:
+        return cls(
+            default,
+            cls.parse_visibility_rules(all),
+            FrozenDict(
+                {
+                    type_alias: cls.parse_visibility_rules(rules)
+                    for type_alias, rules in targets.items()
+                }
+            ),
+        )
+
+    @classmethod
+    def parse_visibility_rules(cls, rules: Iterable[str | DependencyRule]) -> DependencyRules:
```

Comment:
When would we pass DependencyRule, instead of str, here?

Reply:
This method is called from `create()`, for the `all` and `targets.keys()` values which are `Iterable[str | DependencyRule]` due to the `BuildFileDependencyRulesParserState.set_dependency_rules()` method that may inherit rules from a `parent`, where the rules will have been parsed already, so the rules in the rules parser state is `str | DependencyRule`.

Reply:
Gotcha, this is a little confusing, but not horrible.